### ISSUE TITLE
[Bug] IconLayer will sometimes not redraw when rendered as SubLayer of TileLayer

### ISSUE THREAD

**Issue by jeffpeck10x** [visgl/deck.gl#6981]
### Description

I have a TileLayer inside a CompositeLayer that has a `renderSubLayers` prop like this:
```javascript
    const renderSubLayers = ({ data, id }) => {
      return new IconLayer(
        this.getSubLayerProps({
          data,
          filled: true,
          getColor: () => [255, 255, 255],
          getIcon: () => "marker",
          iconAtlas:
            "https://raw.githubusercontent.com/visgl/deck.gl-data/master/website/icon-atlas.png",
          iconMapping: ICON_MAPPING,
          id: `icon-layer-${id}`,
          sizeMinPixels: 10,
          updateTriggers: {},
        })
      );
```

with:
```javascript
const ICON_MAPPING = {
  marker: { x: 0, y: 0, width: 128, height: 128, mask: true },
};
```
(from the deck.gl example).

The CompositeLayer returns the `TileLayer` with that `renderSubLayers` here, like this:
```javascript
    return [
      new TileLayer({
        data,
        extent,
        getTileData,
        maxCacheByteSize,
        maxCacheSize,
        maxRequests,
        maxZoom,
        minZoom,
        onTileLoad,
        refinementStrategy,
        renderSubLayers,
        tileSize,
        updateTriggers: {
          getTileData: updateTriggers.getTileData,
        },
        zoomOffset,
      }),
    ];
```

When I load the application, it will render some of the tiles. If I hover the mouse cursor over any part of the viewport and move it, it will instantly draw the rest of the tiles.

Now, if I zoom or pan, it will load and render tiles as expected, although if I stop moving the mouse cursor, it will periodically not render some of the tiles.

---

I attempted a small workaround - to defer a "needs redraw" trigger to the CompositeLayer just before `return new IconLayer` in `renderSubLayers`:
```javascript
    const renderSubLayers = ({ data, id }) => {
      setTimeout(() => this.setNeedsRedraw(true), 0); // <-- added this line
      return new IconLayer(
        ...
```

Doing so seems to "fix" the problem of tiles not drawing, but I am then left with a problem of older tiles mysteriously disappearing before new tiles are drawn.

---

To contrast, I made a second `renderSubLayers` function that uses the `ScatterplotLayer`:
```javascript
    const renderSubLayersB = ({ data, id }) => {
      return new ScatterplotLayer(
        this.getSubLayerProps({
          data,
          filled: true,
          getFillColor: () => [235, 186, 52, 200],
          getLineColor: () => [52, 79, 235],
          getPosition: (d: any) => d.position,
          id: `scatterplot-layer-${id}`,
          lineWidthMaxPixels: 1,
          lineWidthMinPixels: 0.2,
          opacity: 0.8,
          pickable: true,
          radiusMaxPixels: 8,
          radiusMinPixels: 2,
          radiusScale: 3,
          stroked: true,
        })
      );
    }
```

When I use the ScatterplotLayer, there is no such issue. Everything renders perfectly. There is no issue of tiles randomly not drawing. And there is no issue of old tiles disappearing before new tiles are drawn. It's really quite awesome! I just would like to get to the bottom of why the IconLayer does not work as smoothly.

---

There seems to be a race-condition about when the IconLayer is truly done rendering that does not happen in the ScatterplotLayer. I presume it to be related the asynchronous loading of the icons that needs to happen before anything is rendered. This is all just a guess though. I am not clear if this is an actual bug in the IconLayer or just something that I need to set the appropriate props. Either way, I would really appreciate any guidance on this.

I am using deck.gl `8.6.9` at the moment.

### Flavors

- [x] React
- [ ] Python/Jupyter notebook
- [ ] MapboxLayer
- [ ] GoogleMapsOverlay
- [ ] CartoLayer
- [ ] DeckLayer/DeckRenderer for ArcGIS

### Expected Behavior

* That the TileLayer will draw the IconLayer tiles without the need to move the mouse cursor inside the viewport or have to call:
```
setTimeout(() => this.setNeedsRedraw(true), 0);
```
* That the IconLayer will render/draw as smoothly as the ScatterplotLayer does inside the TileLayer

### Steps to Reproduce

I could try to put together a small example of this, but I am hoping that the above description gives enough information.

### Environment

- Framework version: 8.6.9
- Browser: Chrome 102.0.5005.61
- OS: MacOS Big Sur 11.6.6

**Comment by Pessimistress** [comment]
I recommend you create a Codepen/Code Sandbox to help with the conversation.

I noticed that your TileLayer constructor does not define an id. If you have multiple instances of your custom composite layer that could lead to unexpected behaviors.

**Comment by jeffpeck10x** [comment]
Thanks for the fast response and noticing the lack of id. I tried that (and wrapping with this.getSubLayerProps) although it did make a difference.

I actually went ahead and made a sample project, hoping to reproduce the issue:
https://github.com/jeffpeck10x/deckgl-tilelayer-issue

Although that sample project is working perfectly! So, I guess I need to find the boundary of where my actual application differs from the example to see if is really a bug, or simply something that I can be doing differently.

---

One idea that I had is that the IconLayer calls `setNeedsRedraw` after the icons have loaded. So, what if that gets called _after_ `renderSubLayers` has been called by the TileLayer and the full redraw is triggered.

Like I noted above, it _will_ redraw if I move the cursor over the viewport. So, it's like it needs something to trigger that redraw again.

I'm not sure if any of that makes sense, but I'm just trying to make sense of it.

If you have any ideas to try, I'd appreciate that. I will just stay at it otherwise. Fyi, I plan to be away for the next three days, so I won't be responding then, but thank you for the fast response!

**Comment by Pessimistress** [comment]
Redraw flags are checked every animation frame, so the timing should not matter. You can try set a break point here: 

https://github.com/visgl/deck.gl/blob/8.6-release/modules/core/src/lib/layer-manager.js#L113

And see 1) if the icon layer is in the list; 2) if it is reporting the dirty flag correctly.

**Comment by jeffpeck10x** [comment]
I paused at that breakpoint when some of the IconLayer's were not rendering (it happens every time in my application, until I hover the mouse somewhere and then they appear.)

This is what I found:
<img width="1391" alt="Screen Shot 2022-06-07 at 12 36 02 AM" src="https://user-images.githubusercontent.com/75277973/172296872-b85c063a-4b52-48ec-9b39-508781fbbd77.png">

The three "Initialized" IconLayers at positions 2-0, 2-1, 2-2 seem to correspond to the locations of the IconLayers that are not rendering.

In all cases `needsRedraw` is false.

Even if I deactivate the breakpoint and then reactivate it, they remain in that `Initialized` state and never redraw.

Now, if I deactivate the breakpoint and simply move my mouse cursor anywhere over the viewport, those tiles instantly are drawn and if I then reactivate the breakpoint and view their lifecycle, it says `Matched. State transferred from previous layer` just like the others.

**Comment by jeffpeck10x** [comment]
Well, I figured out what was causing the error after comparing the example where it works with the actual application where it does not work.

I was wrapping the IconLayer's props in `this.getSubLayerProps`.  When I take that away, everything works. (I still have it on the outer TileLayer - just not the IconLayer subLayer)

I am not clear what that is doing that would cause this strange behavior, but I am happy that this is resolved.

I will leave this open in case there is anything else that ought to be investigated with this.

**Comment by Pessimistress** [comment]
Can you reproduce this in your test repo?

**Comment by jeffpeck10x** [comment]
Ah, good idea. In fact, yes, I can!

https://user-images.githubusercontent.com/75277973/172436891-19062969-3f24-416b-894c-84f0f3dcf8fd.mov

In this video, it seems like the action needed to get it to redraw is to actually click and drag (i.e. pan). Whereas in my application, even simply hovering over anything causes the redraw. I am not sure why there is that subtle difference, but either way, this demonstrates that as you pan around, tiles will be randomly left undrawn and only will redraw once you start panning again.

If you remove the `this.getSubLayerProps` on line 65 of `app.js`, the problem goes away and everything redraws as expected.

**Comment by jeffpeck10x** [comment]
Here is a much more pronounced example of this happening:
https://user-images.githubusercontent.com/75277973/172438828-83b0dd23-025a-4bfb-ba55-eae716744f3a.mov

**Comment by Pessimistress** [comment]
To disable the default `data` prop behavior, you overwrite the prop type by supplying a `MyCompositeLayer.defaultProps` object, e.g. https://github.com/visgl/deck.gl/blob/8.7-release/modules/geo-layers/src/tile-layer/tile-layer.js#L8

If you just need to pass the props through to a TileLayer, you can extend the TileLayer instead.

**Comment by jeffpeck10x** [comment]
You know... it didn't even occur to me to extend TileLayer, so thank you for that suggestion.

And I am not clear as to how supplying a default of `undefined` for `defaultProps` would make a difference, although I can try that.

Thank you for the suggestions. The one thing that I would like to understand is how my setting `props.fetch = undefined` caused that strange behavior at all. I would think that if the IconLayer needs to use that `fetch` function and it is `undefined`, it would simply fail. But, it _does_ fetch things. It just randomly doesn't redraw.

I still would regard this as a bug, even if it is a very unlikely one and only due to what seems to be my misuse, by attempting to unset a prop in the constructor.

**Comment by Pessimistress** [comment]
Here is the defaultProps definition for `data` on the Layer class:

`{type: 'data', value: EMPTY_ARRAY, async: true}`

`async` tells the layer that this resource may need to be loaded from an URL.

Here is the defaultProps definition for `data` on the TileLayer class:

`[]`

By removing `async: true` the layer no longer try to fetch it.

You can read about prop definitions in https://deck.gl/docs/developer-guide/custom-layers/prop-types

Here is the definition for `iconAtlas`:

`{type: 'image', value: null, async: true}`

If `props.fetch` is `undefined`, then no "fetching" happens and the prop value resolves to the URL string itself. Because the prop type is `image`, the layer then passes the string as data to `Texture2D`, where the image does get loaded. However, now that some code path outside of the layer is doing the loading, the layer does not know to redraw when loading completes.

TL;DR textures are weird, missing `fetch` prop will not work at all on an other prop type.

**Comment by jeffpeck10x** [comment]
Thank you for explaining that so clearly.

I now see that the important prop to enable/disble fetching is `async`.

And I totally understand why the icons don't load right away if fetch is not defined.

I wonder if it would make sense to check for this scenario and don't try to create the Texture2D if the `fetch` either fails or is undefined for any reason. It seems like deckgl should block the scenario where `Texture2D` does some default behavior that is outside of what the user expects.

**Comment by Pessimistress** [comment]
The proposed behavior will be in v8.8.0
